### ISSUE TITLE
SVG animations of games

### ISSUE THREAD

**Issue by rhaps0dy** [sotetsuk/pgx#739]
(Would you accept a PR like this if I made it?)

It is possible to make SVG animations using the technique in [svgasm](https://github.com/tomkwok/svgasm/blob/master/src/svgasm.cpp). The basic idea is to write a file with 1 group per frame, which looks roughly like the following:

```html
<svg xmlns="http://www.w3.org/2000/svg">
  <defs>
    <style type="text/css">#_0,#_1,#_2,#_3{visibility:hidden}
    </style>
  </defs>
  <g id="_0"> ... frame 0 svg ... </g>
  <g id="_1"> ... frame 1 svg ... </g>
  <g id="_2"> ... frame 2 svg ... </g>
  <g id="_3"> ... frame 3 svg ... </g>
  <defs>
    <style type="text/css">#_0,#_1,#_2,#_3{animation:4s linear _k infinite}#_0{animation-delay:0s}#_1{animation-delay:1s}#_2{animation-delay:2s}#_3{animation-delay:3s}@keyframes _k{0%,25.000000%{visibility:visible}25.000025%,100%{visibility:hidden}}#_{visibility:hidden}
    </style>
  </defs>
</svg>
```
Here each frame lasts for 1 second, but it's easy to make it last for a different amount -- just change the numbers in `animation-delay:`.

The API for this could look like the following:

```python
class State:
    ...
    def save_svg(
        self,
        filename,
        *,
        color_theme: Optional[Literal["light", "dark"]] = None,
        scale: Optional[float] = None,
        time_axis: Optional[int] = None,
        animation_frame_seconds: float = 0.2,
    ) -> None:
        """
        color_theme: Default(None) is "light"
        scale: change image size. Default(None) is 1.0
        time_axis: if not None, animate the state along the indicated axis.
        """
        ...
```

**Comment by sotetsuk** [comment]
Thank you for raising this issue! 
It sounds super cool 👍 We will definitely merge it if you kindly send the PR!

My concern about this feature includes:

1. **Can this approach work with Chess/Mahjong?** Currently, our Chess/Mahjong visualization uses external SVG materials (https://github.com/sotetsuk/pgx/tree/main/pgx/_dwg/images). I'm interested in whether `svgsam` can deal with these environments.
    - if not, we may prefer gif format? as it is more popular for animation. We support svg mainly for academic paper writing.
3. Multi-state support.
4. **On which device, `svgasm` works?** Currently, we limit our dependency as small as possible (https://github.com/sotetsuk/pgx/blob/main/setup.py#L15).


As for the API, we may prefer different realization since our state does not include the past state history.
I may prefer `pgx.save_svg(states: State | Sequence[State], ...)`. So you may test your implementation by the following usage (we will release Chess env in the next release):


```py
import pgx

env = pgx.make("chess")
init = jax.jit(jax.vmap(env.init))
step = jax.jit(jax.vmap(env.step))

batch_size = 4
rng = jax.random.PRNGKey(9999)

states = []
for _ in range(num):
    rng, subkey = jax.random.split(rng)
    keys = jax.random.split(subkey, batch_size)
    state = init(keys)
    states.append(state)
    while not state.terminated.all():
        rng, subkey = jax.random.split(rng)
        action = act_randomly(subkey, state)
        state = step(state, action)
        states.append(state)

pgx.save_svg(states, "states.svg")  # save svg animation
pgx.save_svg(state[0], "first_state.svg")  # save the svg of initial state
pgx.save_svg(state[-1], "first_state.svg")  # save the svg of terminal state
```

Anyway, I'm happy to review and merge your PR! 👍 @rhaps0dy

**Comment by sotetsuk** [comment]
I may have misunderstood the intention of this PR.

**Comment by rhaps0dy** [comment]
No I think you understood it well. But I would copy the functionality of `svgasm` (it's quite small) without including it.

**Comment by rhaps0dy** [comment]
I also quite like the `Sequence[State]` API, it can re-use the existing grid formatting code without much else to do!

* Chess/Mahjong: I think it should work. It works with arbitrary SVGs.
* Multi-state: That's why I thought a "time axis" would work nicely, because if you have a multi-State with multiple leading axes then one of them could be time.
* Which device: it's CPU string manipulation, just like the current SVG drawing code.

**Comment by sotetsuk** [comment]
> No I think you understood it well. But I would copy the functionality of `svgasm` (it's quite small) without including it.

Great! 👍 

You may be able to start adding your preliminary codes to `pgx/_visualizer.py` or `pgx/experimental/xxx.py`.
Feel free to send your PR even if it's not ready (WIP). @OkanoShinri and I can help 👍
